Merge GLSL and HLSL fragment shader generators into one generic

hrydgard opened the pull request:
More refactoring. Finally we'll be able to make fragment shader changes in one single place. 

It has gotten a little bit more complex, but really not that bad.

Vertex shader generators coming up next.

hrydgard committed:
HLSL: Use #define to switch to glsl-like vector types.

hrydgard committed:
wip

hrydgard committed:
Start adding HLSL support to the GLSL shader generator.

hrydgard committed:
Make discard work the same way in the HLSL shader generator.

hrydgard committed:
Change fragment shader output to work the same in GLSL and HLSL.

hrydgard committed:
Evolve the HLSL and GLSL fragment shader generators even closer together.

hrydgard committed:
Set up the test for D3D9, start fixing stuff.

hrydgard committed:
More D3D9 fixes.

hrydgard committed:
HLSL fix

hrydgard committed:
Get D3D9 fragment shader generation to parity

hrydgard committed:
Test parsing of generated OpenGL shaders too (by using glslang).

hrydgard committed:
Minor D3D9 fixes

hrydgard committed:
Misc cleanup

hrydgard committed:
Switch D3D11 over to use the GLSL fragment shader generator

hrydgard committed:
Switch D3D9 too to the glsl shader generator

hrydgard committed:
Remove the HLSL fragment shader generator. The GLSL one can now do its job.

hrydgard committed:
Check for a couple additional GL extensions

hrydgard committed:
Cleanup

hrydgard committed:
Fragment shader gen: Remove unnecessary allocations

hrydgard committed:
Just some cleanups that get to tag along

hrydgard committed:
Move the now-renamed fragment shader generator to GPU/Common.

aliaspider commented on GPU/Common/FragmentShaderGenerator.cpp:
```diff
@@ -283,22 +445,30 @@ bool GenerateFragmentShaderGLSL(const FShaderID &id, char *buffer, const GLSLSha
 	if (isModeClear) {
 		// Clear mode does not allow any fancy shading.
-		WRITE(p, "  vec4 v = v_color0;\n");
+		WRITE(p, "  vec4 v = %sv_color0;\n", compat.inPrefix);
```
wouldn't it be simpler to define those directly under the `struct PS_IN` definition ?

hrydgard replied:
Ah, yeah, that would also be an option. Cleaner indeed, slightly more work for the shader optimizer but should be negligible.

hrydgard replied:
Oh right you mean the preprocessor define method, not declaring new variables in main. Yes actually that was the first thing I, but couldn't get it to work - some weird interaction with HLSL's preprocessor, I dunno. I'm gonna try again.

hrydgard replied:
Yeah, still not working - maybe you have an explanation for this?

```
   1:  #define vec2 float2
   2:  #define vec3 float3
   3:  #define vec4 float4
   4:  #define uvec3 uint3
   5:  #define ivec3 int3
   6:  #define splat3(x) float3(x, x, x)
   7:  #define mix lerp
   8:  #define mod(x, y) fmod(x, y)
   9:  #define DISCARD clip(-1)
  10:  #define DISCARD_BELOW(x) clip(x)
  11:  sampler tex : register(s0);
  12:  vec4 u_texclamp : register(c8);
  13:  vec4 u_alphacolorref : register(c1);
  14:  vec4 u_alphacolormask : register(c2);
  15:  vec3 roundAndScaleTo255v(float3 x) { return floor(x * 255.0f + 0.5f); }
  16:  struct PS_IN {
  17:    vec3 v_texcoord: TEXCOORD0;
  18:    vec4 v_color0: COLOR0;
  19:  };
  20:  #define v_color0 In.v_color0
  21:  #define v_color1 In.v_color1
  22:  #define v_texcoord In.v_texcoord
  23:  #define v_fogdepth In.v_fogdepth
  24:  vec4 main( PS_IN In ) : COLOR {
  25:    vec4 v = v_color0;
  26:    v.a = 0.0;
  27:    v.rgb = splat3(1.0) - v.rgb;
  28:    return v;
  29:  }
C:\dev\ppsspp\unittest\Shader@0x0000020389623FB0(25,12-19): error X3004: undeclared identifier 'v_color0'

```

aliaspider replied:
oh that's weird -.- maybe it didn't like the recursion in the macro or something ...

hrydgard replied:
I'm gonna do it with variable assignment instead, can't be bothered digging deeper into that..

aliaspider replied:
yeah it was the recursion, it works like this : 
```
struct PS_IN {
  vec3 texcoord: TEXCOORD0;
  vec4 color0: COLOR0;
};
#define v_color0 In.color0
#define v_color1 In.color1
#define v_texcoord In.texcoord
#define v_fogdepth In.fogdepth
```

hrydgard replied:
Scary with recursive defines, heh. I'll stick with the variables for now, works fine.

aliaspider replied:
it should have just ignored the recursion like any other C preprocessor, but no :/

hrydgard committed:
UWP buildfix

hrydgard committed:
Fix silly issue in vertex shader gen

hrydgard committed:
Remove inPrefix. HLSL shaders gets slightly longer but probably no biggie.

The extra variable should be optimized out anyway.

Tried doing it with the preprocessor, couldn't get it to work.
